ledger/participant-integration-api/src/main/scala/platform/apiserver/error/TransactionError.scala

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.error
+
+import com.daml.error.ErrorCode.truncateResourceForTransport
+import com.daml.error.{BaseError, ErrorCode}
+import com.daml.ledger.participant.state.v2.Update.CommandRejected.{
+  FinalReason,
+  RejectionReasonTemplate,
+}
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.google.rpc.status.{Status => RpcStatus}
+
+trait TransactionError extends BaseError {
+  def createRejection(
+      correlationId: Option[String]
+  )(implicit
+      logger: ContextualizedLogger,
+      loggingContext: LoggingContext,
+  ): RejectionReasonTemplate = {
+    FinalReason(rpcStatus(correlationId))
+  }
+
+  // Determines the value of the `definite_answer` key in the error details
+  def definiteAnswer: Boolean = false
+
+  final override def definiteAnswerO: Option[Boolean] = Some(definiteAnswer)
+
+  def rpcStatus(
+      correlationId: Option[String]
+  )(implicit logger: ContextualizedLogger, loggingContext: LoggingContext): RpcStatus = {
+
+    // yes, this is a horrible duplication of ErrorCode.asGrpcError. why? because
+    // scalapb does not really support grpc rich errors. there is literally no method
+    // that supports turning scala com.google.rpc.status.Status into java com.google.rpc.Status
+    // objects. however, the sync-api uses the scala variant whereas we have to return StatusRuntimeExceptions.
+    // therefore, we have to compose the status code a second time here ...
+    // the ideal fix would be to extend scalapb accordingly ...
+    val ErrorCode.StatusInfo(codeGrpc, messageWithoutContext, contextMap, _) =
+      code.getStatusInfo(this, correlationId, logger)(loggingContext)
+
+    // TODO error codes: avoid appending the context to the description. right now, we need to do that as the ledger api server is throwing away any error details
+    val message =
+      if (code.category.securitySensitive) messageWithoutContext
+      else messageWithoutContext + "; " + code.formatContextAsString(contextMap)
+
+    val definiteAnswerKey =
+      "definite_answer" // TODO error codes: Can we use a constant from some upstream class?
+
+    // TODO error codes: ensure we don't exceed HTTP header size limits
```

Comment:
this should be fine now. i think it's a stray todo.

Reply:
Removed